### DISCUSSION TITLE
소스코드의 평가, 실행 과정을 설명해주세요

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#74] [실행컨텍스트]
> 전역 코드의 평가, 실행 과정을 설명해주세요.
> 함수 코드의 평가, 실행 과정을 설명해주세요

**Comment by pipisebastian** [comment]
> 전역 코드의 평가, 실행 과정을 설명해주세요.
- **평가**
    - 전역 실행 컨텍스트 생성
    - `환경 Record` 에 식별자 정보를 저장
        - `객체 환경 Record` : var 키워드 변수, 함수 선언문, 빌트인 전역 프로퍼티/함수
            - `BindingObject` 객체와 연결되는데, 이는 전역객체다.
            - ⇒ 그렇기 때문에, var로 선언한 변수같은 경우, 전역 객체의 프로퍼티로 들어가지게 되며.
            - ⇒ 전역객체 window 로 접근을 하지 않더라고, 프로퍼티를 참조할 수 있게 된다.
        - `선언적 환경 Record` :  let, const 키워드 변수
            - let, const 키워드 변수가 들어가는데, 선언과 초기화단계가 분리되므로, 선언문 이전에는 `uninitailzed` 상태를 가집니다.
    - this 바인딩 결정 → `GlobalthisValue` 슬롯에 저장 (전역코드에서는 전역객체)
    - `outerEnvironment Reference` 에는 null 할당.

- **실행 :** 선언문을 제외한 코드가 한줄씩 실행
    - 값의 재할당이 일어남에 따라 실행컨텍스트의 `환경 Record`를 다시 업데이트
    - 식별자를 참조하기 위해 실행컨텍스트에서 검색


> 함수 코드의 평가, 실행 과정을 설명해주세요.

- **평가**
    - 함수 실행 컨텍스트 생성
    - `enviroment Record` 에 식별자 정보를 저장
        - 매개변수, 지역변수, **arguments 객체**
    - this 바인딩 결정 → `thisValue` 슬롯에 저장 (함수로써 호출할때는 전역 객체)
    - 실행컨텍스트→`lexical enviroment`→`enviroment Record`에 식별자 정보가 저장됩니다.
- **실행**
  - 전역코드와 동일 (매개변수, 지역변수에 값 할당)

**Comment by publdaze** [comment]
> 전역 코드의 평가, 실행 과정을 설명해주세요.
### 평가
- 전역 코드가 평가되기 전 전역 객체가 생성됨
- 전역 코드가 평가되기 시작하면 전역 실행 컨텍스트가 생성되고, 실행 컨텍스트 스택에 푸시 됨
- 그리고 전역 렉시컬 환경이 생성되고 전역 실행 컨텍스트에 바인딩 됨
- 전역 렉시컬 환경에는 환경 레코드와 외부 렉시컬 환경에 대한 참조로 이루어지는데
- 환경 레코드에는 전역 환경 레코드의 경우 객체 환경 레코드와 선언적 환경 레코드로 나눠져 바인딩되는데,
- 객체 환경 레코드에는 var로 선언된 변수와 함수선언문이 관리되는데, 평가전 생성된 전역 객체(BindingObject)가 바인딩 되어 있게 때문에 var로 선언된 변수와 함수선언문은 전역 객체의 프로퍼티와 메서드가 됨
- 선언적 환경 레코드는 es6부터 도입된 let, const로 선언된 변수가 등록되어 관리됨
- let, const는 블록 레벨 스코프이기 때문에 분리되어 전역 객체에 들어가지 않고 따로 관리되는 것
- 또 전역 환경 레코드에는 내부 슬롯인 [[GlobalThisValue]]가 존재하는데 이 슬롯에는 this가 바인딩 됨, 일반적으로 전역 코드의 this는 전역 객체를 가리키기 때문에 [[GlobalThisValue]] 내부 슬롯에는 전역 객체가 바인딩 됨
- 외부 렉시컬 환경에 대한 참조는 상위 렉시컬 환경이 참조되며, 식별자를 검색해 나가는 과정에 현재 렉시컬 환경에 식별자가 존재하지 않을 경우 상위 렉시컬 환경에서 탐색해 나가는 과정을 수행하며 스코프 체인이 동작하는데 주요한 역할을 함
- 전역 렉시컬 환경은 스코프 체인의 종점에 존재하므로 전역 렉시컬 환경의 외부 렉시컬 환경에 대한 참조는 null이 됨

### 실행
- 전역 코드가 실행되면서 변수 할당문과 함수 호출문을 만나면 해당 식별자가 존재하는 지를 현재 실행 중인 실행 컨텍스트의 렉시컬 환경부터 존재하는 지 검색을 하는 과정을 거치고 존재하지 않을 경우 외부 렉시컬 환경에 대한 참조에 참조된 상위 렉시컬 환경을 탐색하는 과정을 반복하며 식별자가 존재할 경우 식별자가 결정이 됨
- 변수 할당문은 해당 식별자에 새로운 할당된 값을 바인딩하고
- 함수 호출문은 해당 함수 식별자에 바인딩된 함수를 호출함
- 함수가 호출되면 전역 코드의 실행을 일시 중단하고 함수 코드를 평가하기 시작하고 함수의 평가와 실행 과정이 끝나게 되면 다시 전역 코드를 이어서 진행함.

> 함수 코드의 평가, 실행 과정을 설명해주세요
### 평가
- 함수 코드가 평가될 때 함수 실행 컨텍스트가 생성이 되고, 렉시컬 환경이 생성된 후 실행 컨텍스트 스택에 푸시됨
- 함수의 렉시컬 환경은 환경 레코드와 외부 렉시컬 환경으로 이루어져있음
- 이때 함수의 환경 레코드는 전역 환경 레코드 처럼 객체 환경 레코드와 선언적 환경 레코드로 나뉘지 않고, 환경 레코드 하나로만 이루어져있음
- 이 환경 레코드에는 매개변수, 지역 변수, arguments 객체가 등록되어 관리 됨
- 또 환경 레코드의 내부 슬롯으로 [[ThisValue]]가 존재하는데 이 내부 슬롯에 this가 바인딩되고, 일반적인 함수는 전역 객체가 this로 바인딩 됨
- 외부 렉시컬 환경에 대한 참조에는 상위 렉시컬 환경, 즉 함수 정의가 평가된 시점에 실행중인 실행 컨텍스트의 렉시컬 환경의 참조가 할당됨
### 실행
- 함수가 실행되면 매개변수에 인수가 할당되고, 지역변수의 변수 할당문을 만났을 때 식별자 결정 과정을 거치고 식별자가 존재할 경우 할당됨, 함수 호출문을 만나면 동작을 일시 중단하고 하위 함수의 함수 평가, 실행 과정을 거치고 끝나면 이어서 진행
